In [64]:
import pprint
import sqlite3
import pandas as pd
from IPython.display import display

import os
from hurry.filesize import size


conn = sqlite3.connect('swlondon.sqlite')
cur = conn.cursor()

In [65]:
# ONLY RUN ONCE PER SESSION!!

dirpath = '.'

file_size_wanted = ['nodes.csv','nodes_tags.csv','swlondon.osm','swlondon.sqlite',
                    'ways.csv','ways_nodes.csv','ways_tags.csv']

files_list = []
for path, dirs, files in os.walk(dirpath):
    files_list.extend([(filename, size(os.path.getsize(os.path.join(path, filename)))) for filename in files])

    

In [66]:
files_list

[('.DS_Store', '6K'),
 ('csv_to_database.py', '4K'),
 ('data_wrangling_schema.sql', '833B'),
 ('element_type_count.py', '643B'),
 ('initial_street_name_audit.py', '1K'),
 ('k_attrib_audit.py', '626B'),
 ('nodes.csv', '85M'),
 ('nodes_tags.csv', '5M'),
 ('osm_db_queries.ipynb', '97K'),
 ('osm_export_sg.jpeg', '714K'),
 ('osm_processing_code.ipynb', '28K'),
 ('osm_project_code.ipynb', '244K'),
 ('osm_project_notes.ipynb', '38K'),
 ('OSM_project_submission.ipynb', '21K'),
 ('osm_sampling.py', '948B'),
 ('postcode_audit.py', '2K'),
 ('process_to_csv.py', '12K'),
 ('schema.py', '2K'),
 ('schema.pyc', '1K'),
 ('street_name_audit.py', '3K'),
 ('swlondon.osm', '249M'),
 ('swlondon.sqlite', '130M'),
 ('swlondon_sample.osm', '25M'),
 ('swlondon_sample_submission.osm', '10M'),
 ('tag_type_review.py', '1K'),
 ('ways.csv', '9M'),
 ('ways_nodes.csv', '32M'),
 ('ways_tags.csv', '15M'),
 ('osm_db_queries-checkpoint.ipynb', '56K'),
 ('osm_processing_code-checkpoint.ipynb', '27K'),
 ('osm_project_code-c

In [67]:
files_list.sort(key=lambda size: int(size[1].translate(None, "MKB")), reverse=True)

files_list

[('osm_sampling.py', '948B'),
 ('data_wrangling_schema.sql', '833B'),
 ('schema.cpython-36.pyc', '810B'),
 ('osm_export_sg.jpeg', '714K'),
 ('element_type_count.py', '643B'),
 ('k_attrib_audit.py', '626B'),
 ('osm_project_code-checkpoint.ipynb', '279K'),
 ('swlondon.osm', '249M'),
 ('osm_project_code.ipynb', '244K'),
 ('swlondon.sqlite', '130M'),
 ('osm_db_queries.ipynb', '97K'),
 ('nodes.csv', '85M'),
 ('osm_db_queries-checkpoint.ipynb', '56K'),
 ('osm_project_notes.ipynb', '38K'),
 ('ways_nodes.csv', '32M'),
 ('osm_processing_code.ipynb', '28K'),
 ('osm_project_notes-checkpoint.ipynb', '28K'),
 ('osm_processing_code-checkpoint.ipynb', '27K'),
 ('swlondon_sample.osm', '25M'),
 ('OSM_project_submission.ipynb', '21K'),
 ('OSM_project_submission-checkpoint.ipynb', '19K'),
 ('ways_tags.csv', '15M'),
 ('process_to_csv.py', '12K'),
 ('swlondon_sample_submission.osm', '10M'),
 ('ways.csv', '9M'),
 ('.DS_Store', '6K'),
 ('.DS_Store', '6K'),
 ('nodes_tags.csv', '5M'),
 ('csv_to_database.py', '

In [68]:
file_size_wanted = ['nodes.csv','nodes_tags.csv','swlondon.osm','swlondon.sqlite',
                    'ways.csv','ways_nodes.csv','ways_tags.csv']


for filename, size in files_list:
    if filename in file_size_wanted :
        print '{:.<40s}: {:5s}'.format(filename,size)

swlondon.osm............................: 249M 
swlondon.sqlite.........................: 130M 
nodes.csv...............................: 85M  
ways_nodes.csv..........................: 32M  
ways_tags.csv...........................: 15M  
ways.csv................................: 9M   
nodes_tags.csv..........................: 5M   


In [69]:
df = pd.read_sql_query(
''' 
SELECT * from nodes
LIMIT 3
''' , conn)


print ('nodes table')
display(df)
print  ('\n')


df2 = pd.read_sql_query(
''' 
SELECT * from ways
LIMIT 3
''' , conn)

print ('ways table')
display(df2)
print  ('\n')

df3 = pd.read_sql_query(
''' 
SELECT * from nodes_tags
LIMIT 3
''' , conn)

print ('nodes_tags table')
display(df3)
print  ('\n')


df4 = pd.read_sql_query(
''' 
SELECT * from ways_tags
LIMIT 3
''' , conn)

print ('ways_tags table')
display(df4)
print  ('\n')

df5 = pd.read_sql_query(
''' 
SELECT * from ways_nodes
LIMIT 3
''' , conn)

print ('ways_nodes table')
display(df5)
print  ('\n')

nodes table


id        lat       lon       user     uid  version  changeset  \
0  109642  51.483586 -0.167361  MacLondon  322039       10   43286497   
1  109644  51.482444 -0.172071  Firefishy    3560        5    1873462   
2  109645  51.482151 -0.173407  MacLondon  322039        5   29279984   

              timestamp  
0  2016-10-30T16:09:18Z  
1  2009-07-19T14:38:21Z  
2  2015-03-06T00:39:17Z



ways table


id        user     uid version  changeset             timestamp
0  2987  jpennycook   17292       8   49509164  2017-06-13T19:12:51Z
1  2989       Zorac  501777       2   10852532  2012-03-02T21:16:46Z
2  2991    Wynndale  118262       3   12410104  2012-07-21T13:27:36Z



nodes_tags table


id             key                                              value  \
0  109644  randomjunk_bot  reverted edit by FilipeOliveira on 2008-11-24T...   
1  109645         highway                                    traffic_signals   
2  109645  randomjunk_bot  reverted edit by FilipeOliveira on 2008-11-24T...   

      type  
0  regular  
1  regular  
2  regular



ways_tags table


id      key        value     type
0  2987   access      private  regular
1  2987  highway      service  regular
2  2989  highway  residential  regular



ways_nodes table


id     node_id  position
0  2987  1667684820         0
1  2987  3927098721         1
2  2987  3927098722         2

In [70]:

# Count of Nodes & Ways

df1 = pd.read_sql_query(
''' 
SELECT COUNT(nodes.id) as Nodes
FROM nodes
''' , conn)

print ('Number of Nodes:')
display(df1)
print  ('\n')

df2 = pd.read_sql_query(
''' 
SELECT COUNT(ways.id) as Ways
FROM ways
''' , conn)

print ('Number of Ways:')
display(df2)
print  ('\n')



Number of Nodes:


Nodes
0  1100871



Number of Ways:


Ways
0  174981

In [71]:
# Count unique contributing users

df1 = pd.read_sql_query(
''' 
SELECT COUNT(DISTINCT(all_users.uid)) as UserCount          
FROM (SELECT uid FROM nodes UNION ALL SELECT uid FROM ways) all_users
''' , conn)

df1

UserCount
0       2101

In [72]:
# Count contributions from top 25 contributing users

df = pd.read_sql_query(
''' 
SELECT all_users.user, count(all_users.uid) as contributions
FROM (SELECT uid, user FROM nodes UNION ALL SELECT uid, user FROM ways) as all_users
GROUP BY all_users.uid
ORDER BY Contributions DESC
limit 15;
''' , conn)


df

user  contributions
0            Rondon237         137357
1              ecatmur          94440
2                  OJW          66516
3       joelholdsworth          59131
4          Ian Haylock          47092
5             UniEagle          33680
6               moyogo          31435
7               Mark_S          30999
8             Wynndale          27862
9              Ed Avis          26350
10           marsbar28          24639
11  Paul The Archivist          20251
12      Derick Rethans          19761
13   LivingWithDragons          19489
14        captainpovey          18079

In [73]:
# How many users have only made one contribution?

df = pd.read_sql_query(
''' 
SELECT COUNT(*) as one_contrib_users
FROM
    (SELECT all_users.uid, all_users.user, count(all_users.uid) as contributions
    FROM (SELECT uid, user FROM nodes UNION ALL SELECT uid, user FROM ways) as all_users
    GROUP BY all_users.uid
    HAVING contributions=1) as singles;

''' , conn)


df




one_contrib_users
0                460

In [74]:
df = pd.read_sql_query(
''' 
SELECT value, COUNT(*) as num
FROM nodes_tags
WHERE key='amenity'
GROUP BY value
ORDER BY num DESC
LIMIT 35;

''' , conn)


df

value   num
0           post_box  1607
1    bicycle_parking   792
2         restaurant   663
3              bench   623
4                pub   564
5               cafe   469
6          telephone   438
7          fast_food   321
8                atm   271
9   place_of_worship   271
10      waste_basket   215
11       car_sharing   159
12          pharmacy   155
13              bank   154
14           toilets   150
15         recycling   146
16           parking   132
17       post_office   127
18              fuel   106
19               bar    77
20    bicycle_rental    73
21           dentist    72
22   vending_machine    67
23            school    54
24           doctors    49
25           library    46
26      kindergarten    42
27          hospital    34
28    drinking_water    30
29              taxi    28
30          fountain    26
31            police    22
32  community_centre    21
33   public_building    18
34        veterinary    18

In [75]:
# List of Pub names


df = pd.read_sql_query(
''' 
SELECT pubs.id, pubs.value, names.value
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id
GROUP BY pubs.id

''' , conn)


df

pubs.id pubs.value            names.value
0      13353276        pub             The Harrow
1      17806375        pub              Green Man
2      18235090        pub           The Arab Boy
3      19828422        pub            The Anglers
4      20627020        pub      The Edmund Tylney
5      20692145        pub        The Penny Black
6      20695067        pub             The Plough
7      20973312        pub             Kings Arms
8      20973316        pub          The Waterside
9      21980919        pub          The Manor Inn
10     26108616        pub         Kingswood Arms
11     26422089        pub               The Mint
12     26422124        pub           The Woolpack
13     26461817        pub               The Goat
14     26462174        pub           The Red Lion
15     26545063        pub           The Windmill
16     26545687        pub        The Queens Head
17     26602504        pub           The Feathers
18     27245287        pub   The Shaftesbury Arms
19     27512718        pub              Lots Road
20     30429954        pub           The Tapestry
21     30429955        pub        Jolly Gardeners
22     30441135        pub               The Hope
23     30478182        pub               The Ship
24     30478187        pub          Ye White Hart
25     30483366        pub        The Park Tavern
26     30483782        pub       Alexandra Tavern
27     30512931        pub           The Cardinal
28     30552863        pub          The Greyhound
29     30629850        pub           The Wych Elm
..          ...        ...                    ...
530  3011025198        pub             Kings Arms
531  3014221516        pub               The Bear
532  3068306969        pub             The Anchor
533  3133134050        pub           The Red Lion
534  3247663714        pub        The Park Tavern
535  3282498381        pub          The Kings Tun
536  3340684903        pub              The Crown
537  3362469515        pub       The Magic Garden
538  3423539966        pub         The Earlsfield
539  3523526667        pub         The Two Rivers
540  3598403423        pub         The Anchor Pub
541  3681486426        pub       The Old Red Lion
542  3757136233        pub            The Phoenix
543  3770513926        pub      The Jolly Butcher
544  3782616223        pub                 Steins
545  3793532459        pub           The Yardbird
546  3964556370        pub          The Sands End
547  4026677008        pub        Queen Elisabeth
548  4117016441        pub            The Terrace
549  4203361955        pub      The Prince Albert
550  4436916522        pub       Bricklayers Arms
551  4525422451        pub     The Flying Chariot
552  4527208485        pub       Gorgeous Kitchen
553  4527208488        pub         London's Pride
554  4578415857        pub    The Waldegrave Arms
555  4617068622        pub     Ganley's Irish Bar
556  4629007069        pub      The Running Horse
557  4682459836        pub          The Racehorse
558  4926469145        pub  The Merton Apprentice
559  4928682125        pub         The Cricketers

[560 rows x 3 columns]

In [76]:
#How many distinct pub names?

df = pd.read_sql_query(
''' 
SELECT COUNT(DISTINCT names.value) as num_pub_names
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id


''' , conn)


df

num_pub_names
0            453

In [77]:
# Top 10 Pub Names

df = pd.read_sql_query(
''' 
SELECT names.value, COUNT(names.value) as name_count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id
GROUP BY names.value
ORDER BY name_count DESC
LIMIT 10;

''' , conn)


df

names.value  name_count
0        The Royal Oak           9
1           The Plough           7
2            The Crown           6
3         The Red Lion           6
4           Kings Arms           4
5             The Bell           4
6           The Castle           4
7       The Cricketers           4
8           The George           4
9  The Prince of Wales           4

In [78]:

df = pd.read_sql_query(
''' 
SELECT info.value, count(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='cafe') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type='regular'
    AND key = 'name') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
LIMIT 15;

''' , conn)


df




info.value  count
0            Starbucks     29
1                Costa     21
2           Caffè Nero     15
3           Caffe Nero      5
4        Pret A Manger      5
5         Costa Coffee      4
6                 Nero      4
7     Starbucks Coffee      4
8            Puccino's      3
9        Sunshine Cafe      3
10           Best Cafe      2
11           Cafe Nero      2
12  Caffè Nero Express      2
13          Cuppa-cino      2
14      Harris & Hoole      2

In [79]:
df = pd.read_sql_query(
''' 
SELECT info.value, COUNT(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='car_sharing') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type = 'regular'
    AND key = 'operator') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
;

''' , conn)


df

info.value  count
0      Streetcar    140
1         Zipcar     14
2  City Car Club      1
3      StreetCar      1

In [80]:
df = pd.read_sql_query(
''' 
SELECT info.value, COUNT(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='fuel') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type = 'regular'
    AND key = 'operator') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
;

''' , conn)


df

info.value  count
0                  BP     18
1               Shell     15
2                Esso     10
3         Sainsbury's      4
4               Total      3
5               Tesco      2
6              Texaco      2
7                ESSO      1
8                 Jet      1
9    MRH Hatton Cross      1
10              Murco      1
11  Royal Dutch Shell      1
12         Sainsburys      1
13              tesco      1

In [81]:
# Useful query for looking up info available on different amenity types. 
# just change 'value' in first subquery to the amenity you are looking at

df = pd.read_sql_query(
''' 
SELECT *
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='bicycle_rental') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity') as info

ON amenities.id =  info.id
;

''' , conn)


df

id           value          id       key  \
0     261761004  bicycle_rental   261761004  capacity   
1     780114277  bicycle_rental   780114277      name   
2     823375879  bicycle_rental   823375879     local   
3     823375879  bicycle_rental   823375879      name   
4     823375879  bicycle_rental   823375879   network   
5     823375879  bicycle_rental   823375879       ref   
6     836785671  bicycle_rental   836785671  capacity   
7     836785671  bicycle_rental   836785671     local   
8     836785671  bicycle_rental   836785671      name   
9     836785671  bicycle_rental   836785671   network   
10    836785671  bicycle_rental   836785671       ref   
11    836785673  bicycle_rental   836785673  capacity   
12    836785673  bicycle_rental   836785673     local   
13    836785673  bicycle_rental   836785673      name   
14    836785673  bicycle_rental   836785673   network   
15    836785673  bicycle_rental   836785673       ref   
16    836785680  bicycle_rental   836785680  capacity   
17    836785680  bicycle_rental   836785680     local   
18    836785680  bicycle_rental   836785680      name   
19    836785680  bicycle_rental   836785680   network   
20    836785680  bicycle_rental   836785680       ref   
21    836785688  bicycle_rental   836785688  capacity   
22    836785688  bicycle_rental   836785688     local   
23    836785688  bicycle_rental   836785688      name   
24    836785688  bicycle_rental   836785688   network   
25    836785688  bicycle_rental   836785688       ref   
26    836785695  bicycle_rental   836785695  capacity   
27    836785695  bicycle_rental   836785695     local   
28    836785695  bicycle_rental   836785695      name   
29    836785695  bicycle_rental   836785695   network   
..          ...             ...         ...       ...   
145  3581895443  bicycle_rental  3581895443   network   
146  3581895443  bicycle_rental  3581895443  operator   
147  3750681028  bicycle_rental  3750681028      name   
148  3890101248  bicycle_rental  3890101248      name   
149  3890101249  bicycle_rental  3890101249      name   
150  3890126460  bicycle_rental  3890126460      name   
151  3896102441  bicycle_rental  3896102441      name   
152  4112448142  bicycle_rental  4112448142      name   
153  4112448142  bicycle_rental  4112448142   network   
154  4329814892  bicycle_rental  4329814892      name   
155  4537517403  bicycle_rental  4537517403  capacity   
156  4537517403  bicycle_rental  4537517403      name   
157  4537517403  bicycle_rental  4537517403   network   
158  4537517403  bicycle_rental  4537517403  operator   
159  4547084070  bicycle_rental  4547084070      name   
160  4677086112  bicycle_rental  4677086112  capacity   
161  4677086112  bicycle_rental  4677086112      name   
162  4677086112  bicycle_rental  4677086112   network   
163  4677106589  bicycle_rental  4677106589  capacity   
164  4677106589  bicycle_rental  4677106589      name   
165  4677106589  bicycle_rental  4677106589   network   
166  4677109172  bicycle_rental  4677109172  capacity   
167  4677109172  bicycle_rental  4677109172      name   
168  4677109172  bicycle_rental  4677109172   network   
169  4677113221  bicycle_rental  4677113221  capacity   
170  4677113221  bicycle_rental  4677113221      name   
171  4677113221  bicycle_rental  4677113221   network   
172  4677116099  bicycle_rental  4677116099  capacity   
173  4677116099  bicycle_rental  4677116099      name   
174  4677116099  bicycle_rental  4677116099   network   

                             value     type  
0                                2  regular  
1                        Parkcycle  regular  
2                 South Kensington  regular  
3          Gloucester Road Station  regular  
4                   tfl_cycle_hire  regular  
5                          CH19049  regular  
6                               18  regular  
7                     Earl's Court  regular  
8                    Trebovir Road  regular  
9                 

post_box	1607
1	bicycle_parking	792
2	restaurant	663
3	bench	623
4	pub	564
5	cafe	469
6	telephone	438
7	fast_food	321
8	atm	271
9	place_of_worship	271
10	waste_basket	215
11	car_sharing	159
12	pharmacy	155
13	bank	154
14	toilets	150
15	recycling	146
16	parking	132
17	post_office	127
18	fuel	106
19	bar	77
20	bicycle_rental	73
21	dentist	72
22	vending_machine	67
23	school	54
24	doctors	49